File: backend/src/models/property.py
from datetime import datetime, timezone
import uuid
from typing import Any
from geoalchemy2 import Geometry
from pgvector.sqlalchemy import Vector
from sqlalchemy import (
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    func,
    inspect,
    text,
)
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.core.config import settings
from src.core.database import Base


class Property(Base):
    __tablename__ = "properties"
    __table_args__ = (
        Index(
            "ix_properties_embedding_hnsw",
            "embedding",
            postgresql_using="hnsw",
            postgresql_with={"m": 16, "ef_construction": 64},
            postgresql_ops={"embedding": "vector_cosine_ops"},
        ),
        Index(
            "ix_properties_fts",
            text(
                "to_tsvector('simple', coalesce(title, '') || ' ' || coalesce(address, '') || ' ' || coalesce(ward, '') || ' ' || coalesce(district, '') || ' ' || coalesce(city, '') || ' ' || coalesce(description, ''))"
            ),
            postgresql_using="gin",
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        index=True,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    property_type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    listing_type: Mapped[str] = mapped_column(String(20), nullable=False, index=True)  # sale / rent
    price: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, index=True)
    currency: Mapped[str] = mapped_column(String(10), nullable=False, default="VND")
    area_sqm: Mapped[float] = mapped_column(Float, nullable=False, index=True)
    num_bedrooms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    num_bathrooms: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # Location
    address: Mapped[str] = mapped_column(String(500), nullable=False)
    ward: Mapped[str | None] = mapped_column(String(100), nullable=True)
    district: Mapped[str | None] = mapped_column(String(100), nullable=True, index=True)
    city: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    geom: Mapped[Any | None] = mapped_column(
        Geometry(geometry_type="POINT", srid=4326, spatial_index=True),
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="active", server_default="active", index=True
    )

    # Owner user id
    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    owner = relationship("User", back_populates="properties")

    # Parent project id
    project_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    project = relationship("Project", back_populates="properties", lazy="selectin")

    # Listing images
    images: Mapped[list[str]] = mapped_column(
        ARRAY(Text),
        nullable=False,
        default=list,
        server_default="{}",
    )

    # 768-dimensional vector embedding for semantic search
    embedding = mapped_column(Vector(settings.VECTOR_DIM), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    @property
    def agent(self) -> dict[str, Any] | None:
        state = inspect(self)
        if "owner" in state.dict and self.owner is not None:
            return {
                "id": self.owner.id,
                "full_name": self.owner.full_name,
                "email": self.owner.email,
                "phone_number": self.owner.phone,
                "phone": self.owner.phone,
                "avatar_url": self.owner.avatar_url,
                "role": self.owner.role,
            }
        return None

    def __init__(self, **kwargs):
        if "id" not in kwargs or kwargs["id"] is None:
            kwargs["id"] = uuid.uuid4()
        if "status" not in kwargs or kwargs["status"] is None:
            kwargs["status"] = "active"
        if "currency" not in kwargs or kwargs["currency"] is None:
            kwargs["currency"] = "VND"
        if "images" not in kwargs or kwargs["images"] is None:
            kwargs["images"] = []
        now = datetime.now(timezone.utc)
        if "created_at" not in kwargs or kwargs["created_at"] is None:
            kwargs["created_at"] = now
        if "updated_at" not in kwargs or kwargs["updated_at"] is None:
            kwargs["updated_at"] = now
        super().__init__(**kwargs)

    def __repr__(self) -> str:
        return f"<Property(id={self.id}, title='{self.title[:30]}', listing_type='{self.listing_type}', price={self.price})>"

File: frontend/mobile/lib/models/property.dart
import 'project_models.dart';

class PropertyAgent {
  final String id;
  final String fullName;
  final String email;
  final String? phoneNumber;
  final String? avatarUrl;
  final String role;

  PropertyAgent({
    required this.id,
    required this.fullName,
    required this.email,
    this.phoneNumber,
    this.avatarUrl,
    this.role = 'agent',
  });

  factory PropertyAgent.fromJson(Map<String, dynamic> json) {
    return PropertyAgent(
      id: (json['id'] as String?) ?? '',
      fullName: (json['full_name'] as String?) ?? '',
      email: (json['email'] as String?) ?? '',
      phoneNumber: (json['phone_number'] ?? json['phone']) as String?,
      avatarUrl: json['avatar_url'] as String?,
      role: (json['role'] as String?) ?? 'agent',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'full_name': fullName,
      'email': email,
      'phone_number': phoneNumber,
      'avatar_url': avatarUrl,
      'role': role,
    };
  }
}

class Property {
  final String id;
  final String? userId;
  final String title;
  final String description;
  final String propertyType;
  final String listingType;
  final String? rentalType;
  final Map<String, dynamic>? rentalCosts;
  final Map<String, dynamic>? rentalRules;

  double? get depositAmount {
    final months = rentalCosts?["deposit_months"] as num?;
    return months == null ? null : price * months.toDouble();
  }
  final double price;
  final String currency;
  final double areaSqm;
  final int? numBedrooms;
  final int? numBathrooms;
  final String address;
  final String? ward;
  final String? district;
  final String city;
  final double? latitude;
  final double? longitude;
  final List<String> images;
  final String? projectId;
  final PropertyAgent? agent;
  final ProjectSummary? project;
  final String status;
  final String? createdAt;
  final String? updatedAt;

  Property({
    required this.id,
    this.userId,
    required this.title,
    required this.description,
    required this.propertyType,
    required this.listingType,
    this.rentalType,
    this.rentalCosts,
    this.rentalRules,
    required this.price,
    this.currency = 'VND',
    required this.areaSqm,
    this.numBedrooms,
    this.numBathrooms,
    required this.address,
    this.ward,
    this.district,
    required this.city,
    this.latitude,
    this.longitude,
    this.images = const [],
    this.projectId,
    this.agent,
    this.project,
    required this.status,
    this.createdAt,
    this.updatedAt,
  });

  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: json['id'] as String,
      userId: json['user_id'] as String?,
      title: (json['title'] as String?) ?? '',
      description: (json['description'] as String?) ?? '',
      propertyType: (json['property_type'] as String?) ?? 'apartment',
      listingType: (json['listing_type'] as String?) ?? 'sale',
      rentalType: json['rental_type'] as String?,
      rentalCosts: (json['rental_costs'] as Map<String, dynamic>?),
      rentalRules: (json['rental_rules'] as Map<String, dynamic>?),
      price: ((json['price'] as num?) ?? 0).toDouble(),
      currency: (json['currency'] as String?) ?? 'VND',
      areaSqm: ((json['area_sqm'] as num?) ?? 0).toDouble(),
      numBedrooms: json['num_bedrooms'] as int?,
      numBathrooms: json['num_bathrooms'] as int?,
      address: (json['address'] as String?) ?? '',
      ward: json['ward'] as String?,
      district: json['district'] as String?,
      city: (json['city'] as String?) ?? '',
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      images: (json['images'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      projectId: json['project_id'] as String?,
      agent: json['agent'] != null
          ? PropertyAgent.fromJson(json['agent'] as Map<String, dynamic>)
          : null,
      project: json['project'] != null
          ? ProjectSummary.fromJson(json['project'] as Map<String, dynamic>)
          : null,
      status: (json['status'] as String?) ?? 'active',
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'title': title,
      'description': description,
      'property_type': propertyType,
      'listing_type': listingType,
      'rental_type': rentalType,
      'rental_costs': rentalCosts,
      'rental_rules': rentalRules,
      'price': price,
      'currency': currency,
      'area_sqm': areaSqm,
      'num_bedrooms': numBedrooms,
      'num_bathrooms': numBathrooms,
      'address': address,
      'ward': ward,
      'district': district,
      'city': city,
      'latitude': latitude,
      'longitude': longitude,
      'images': images,
      'project_id': projectId,
      'agent': agent?.toJson(),
      'project': project?.toJson(),
      'status': status,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}
